File: data processing/dataLoader.py
import psycopg2
from psycopg2 import sql
import sys
from config import config
import time

WORKING_PATH = "test-data/"


def connectDatabase():
    # create the connection to the database
    # source code from: https://www.youtube.com/watch?v=Q8iYj2ypWss&list=PLrOQsSoS-V6-LRHG_8q0a4KBkWlTWeKhe&index=4
    connection = None
    try:
        params = config()
        print("Connecting to the postgreSQL database ...")
        connection = psycopg2.connect(**params) # extract everything from params
        # create a cursor
        cursor = connection.cursor()
        print("Connection to PostgreSQL is established!")
        # print the current version of PostgreSQL
        print("PostgreSQL database version:")
        cursor.execute("SELECT version()")
        db_version = cursor.fetchone()
        print(db_version)
    except(Exception, psycopg2.OperationalError) as error:
        print("Could not connect to the database!\n {}".format(error))
        sys.exit(1)

    return connection, cursor


def closeDatabase(connection, cursor):
    # close the cursor
    cursor.close()
    # close connection to database
    connection.close()
    print("Database connection terminated.")


def dropTables(connection, cursor):
    table_names = ["users", "users_info"]
    for table in table_names:
        try:
            cursor.execute(sql.SQL(""" DROP TABLE IF EXISTS {table} CASCADE """).format(table=sql.Identifier(table)))
        except(Exception, psycopg2.DatabaseError) as error:
            print("Could not drop table {table}!\n {error}".format(table=table, error=error))
            connection.rollback()
            cursor.close()
            connection.close()
            sys.exit(1)
    print("Deleted all existing tables successfully!")


def createTables(connection, cursor):
    credentials_query = """ CREATE TABLE IF NOT EXISTS users (
        id SERIAL PRIMARY KEY,
        username VARCHAR(20) UNIQUE NOT NULL,
        roles VARCHAR(20) DEFAULT '2001',
        password VARCHAR(255) NOT NULL,
        refresh_token VARCHAR(255)
    )"""
    users_query = """ CREATE TABLE IF NOT EXISTS users_info (
        username VARCHAR(255) PRIMARY KEY,
        weight NUMERIC DEFAULT NULL,
        height SMALLINT DEFAULT NULL,
        FOREIGN KEY (username) REFERENCES users (username)
    )"""

    query_list = [credentials_query, users_query]

    for query in query_list:
        try:
            cursor.execute(query)
        except(Exception, psycopg2.DatabaseError) as error:
            print("Could not create table {table}!\n {error}".format(table=query.rsplit("_",1), error=error))
            connection.rollback()
            cursor.close()
            connection.close()
            sys.exit(1)

    print("Tables created successfully!")

    # commit the transaction from memory to postgres
    connection.commit()


def loadDataToDB(connection, cursor):
    # define dictionary for CSV file - table name translation
    table_dict = {
        "users.csv":"users",
        "users_info.csv":"users_info"
    }

    for key, value in table_dict.items():
        # open file in read mode
        with open(WORKING_PATH + key, 'r') as f:
            next(f) # skip header row
            try:
                # read data and treat each 'None' value as postgres Null
                # cursor.copy_from(f, value, sep=';', columns=('username','password','weight','height'), null='None')
                cursor.copy_from(f, value, sep=';', null='None')
                connection.commit()
                print("Data from file {} is successfully loaded to table {}".format(key, value))
            except(Exception, psycopg2.DatabaseError) as error:
                print("Could not insert data into table!\n {}".format(error))
                connection.rollback()
                continue


if __name__ == "__main__":
    # create a connection to the postgres database
    connection, cursor = connectDatabase()
    # drop tables if they already exist
    dropTables(connection, cursor)
    # create the tables required
    createTables(connection, cursor)
    # time measure for data loading
    start_time = time.time()
    # load the CSV files into the database
    loadDataToDB(connection, cursor)
    print("Time needed for data loading:\n--- %s seconds ---" % (time.time() - start_time))
    # close the connection to the database
    closeDatabase(connection, cursor)

File: data processing/dataLoader.js

const connectDB = require('./dbConn');
const fs = require("fs");
const fastcsv = require("fast-csv");

const dropDBTable = async (tableName, res) => {
    try {
        const pool = await connectDB();
        await pool.query(`DROP TABLE IF EXISTS ${tableName} CASCADE`);
        console.log(`table ${tableName} dropped successfully`)
    } catch (error) {
        console.log(error.message);
    }
};

const createDBTable = async (tableName,res) => {
    try {
        const pool = await connectDB();
        switch (tableName) {
            case 'users':
                await pool.query(`CREATE TABLE IF NOT EXISTS users (
                    id SERIAL PRIMARY KEY,
                    username VARCHAR(20) UNIQUE NOT NULL,
                    roles VARCHAR(20) DEFAULT '2001',
                    password VARCHAR(255) NOT NULL,
                    refresh_token VARCHAR(255),
                    content JSONB)
                    `);
                break
            case 'spots':
                await pool.query(`CREATE TABLE IF NOT EXISTS spots (
                    spotname VARCHAR(255) PRIMARY KEY,
                    canton VARCHAR(50) DEFAULT NULL,
                    country VARCHAR(50) DEFAULT NULL,
                    latitude NUMERIC DEFAULT NULL,
                    longitude NUMERIC DEFAULT NULL, 
                    content JSONB)
                `);
                break 
            case 'routes':
                await pool.query(`CREATE TABLE IF NOT EXISTS routes (
                    routename VARCHAR(255),
                    spotname VARCHAR(255),
                    grade VARCHAR(3) NOT NULL,
                    content JSONB,
                    PRIMARY KEY (routename, spotname),
                    FOREIGN KEY (spotname) REFERENCES spots (spotname) ON DELETE CASCADE)
                `);
                break 
            // zusammengesetzter Foreign Key nötig damit UNIQUE Constraint erfüllt ist
            case 'user_routes':
                await pool.query(`CREATE TABLE IF NOT EXISTS user_routes (
                    username VARCHAR(255),
                    routename VARCHAR(255),
                    spotname VARCHAR(255),
                    content JSONB,
                    PRIMARY KEY(username, routename),
                    FOREIGN KEY (username) REFERENCES users (username) ON DELETE CASCADE,
                    FOREIGN KEY (routename, spotname) REFERENCES routes (routename, spotname) ON DELETE CASCADE)
                `);
                break
            default:
                console.log('No correct Database found');
                return;
        }
        console.log(`Table ${tableName} created successfully`)
    } catch (error) {
        console.log(error.message);
    }
};

const insertDataToDB = async (data, tableName, res) => {
    try {
        const pool = await connectDB();
        switch (tableName) {
            case 'users':
                var query = `INSERT INTO users (username, roles, password, refresh_token, content) VALUES ($1, $2, $3, $4, $5)`
                break
            case 'spots':
                var query = `INSERT INTO spots (spotname, canton, country, latitude, longitude, content) VALUES ($1, $2, $3, $4, $5, $6)`
                break
            case 'routes':
                var query = `INSERT INTO routes (routename, spotname, grade, content) VALUES ($1, $2, $3, $4)`
                break
            case 'user_routes':
                var query = `INSERT INTO user_routes (username, routename, spotname, content) VALUES ($1, $2, $3, $4)`
                break
            default:
                console.log('No correct Database found');
                return
        }
        data.forEach(row => {
             pool.query(query, row, (error, res) => {
                if (error) {
                    console.log(error.stack);
                };
            });
        });
        console.log(`Data inserted to ${tableName}`);
    } catch (error) {
        console.log(error.message);
    };
};

const runWorkflow = async (tableName) => {
    let stream = await fs.createReadStream(`test-data/${tableName}.csv`);
    let csvData = [];
    let csvStream = await fastcsv.parse({delimiter: ';'})
        .on("data", function(data) {
            csvData.push(data);
        })
        .on("end", function() {
        // remove the first line: header
        csvData.shift();
        // insert data to database
        insertDataToDB(csvData, tableName)
    });
    stream.pipe(csvStream);
};

async function doWork() {
    const tableNames = ["users", "spots", "routes", "user_routes"];
    try {
        for (element of tableNames) {
            await dropDBTable(element);
            await createDBTable(element);
            await runWorkflow(element);
        };
    }
    catch (error) {
        console.log(error.message);
    }
}

doWork();